storage2/src/state.rs

```diff
@@ -1,47 +1,82 @@
-use std::collections::HashMap;
+use std::fmt::Debug;
 
-use jmt::{
-    storage::{TreeReader, TreeWriter},
-    JellyfishMerkleTree,
-};
+use async_trait::async_trait;
+
+use penumbra_proto::{Message, Protobuf};
 
 mod transaction;
-use transaction::Transaction as StateTransaction;
+pub use transaction::Transaction as StateTransaction;
 
 /// State is a lightweight copy-on-write fork of the chain state,
 /// implemented as a RYW cache over a pinned JMT version.
-pub(crate) struct State<'a, R: TreeReader + TreeWriter> {
-    cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
-    jmt_version: jmt::Version,
-    jmt: &'a JellyfishMerkleTree<'a, R>,
+pub struct State {
+    // TODO: determine which fields to include
+    // cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
+    // jmt_version: jmt::Version,
+    // jmt: &'a JellyfishMerkleTree<'a, R>,
 }
 
-impl<'a, R: TreeReader + TreeWriter> State<'a, R> {
+impl State {
     pub fn new() -> Self {
         Self {
-            cache: todo!(),
-            jmt_version: todo!(),
-            jmt: todo!(),
+            // cache: todo!(),
+            // jmt_version: todo!(),
+            // jmt: todo!(),
         }
     }
 
-    pub fn begin_transaction(&mut self) -> StateTransaction<'a, R> {
-        StateTransaction::new()
+    pub fn begin_transaction(&mut self) -> StateTransaction {
+        StateTransaction{
+            // cache: todo!(),
+            // unwritten_changes: todo!(),
+            // state: self,
+        }
+    }
+
+    /// Gets a domain type from the State.
+    pub fn get<D, P>(&self, key: String) -> Option<D>
+    where
+        D: Protobuf<P>,
+        // TODO: does this get less awful if P is an associated type of D?
+        P: Message + Default,
+        P: From<D>,
+        D: TryFrom<P> + Clone + Debug,
+        <D as TryFrom<P>>::Error: Into<anyhow::Error>,
+    {
+        todo!()
+    }
+
+    /// Gets a proto type from the State.
+    pub fn get_proto<D, P>(&self, key: String) -> Option<P>
+    where
+        D: Protobuf<P>,
+        // TODO: does this get less awful if P is an associated type of D?
+        P: Message + Default,
+        P: From<D>,
+        D: TryFrom<P> + Clone + Debug,
+        <D as TryFrom<P>>::Error: Into<anyhow::Error>,
+    {
+        todo!()
```

Comment:
Let's put these as [provided methods](https://doc.rust-lang.org/book/ch10-02-traits.html#default-implementations) on the `StateRead` trait, implemented in terms of the required `get_raw` method, so that they're available for every `StateRead` implementation.